base/staticdata.jl

```diff
@@ -109,175 +165,246 @@ end
+
+                parent_cycle = parent_result.child_cycle
+                parent_minworld = parent_result.result_minworld
+                parent_maxworld = parent_result.result_maxworld
+                parent_cause = parent_work.cause
+                parent_stage = parent_work.stage
+
+                if parent_minworld < min_valid2
+                    parent_minworld = min_valid2
+                end
+                if parent_minworld > max_valid2
+                    max_valid2 = 0
+                end
+                if parent_maxworld > max_valid2
+                    parent_cause = callee
+                    parent_maxworld = max_valid2
+                end
+                if max_valid2 == 0
+                    # found what we were looking for, so terminate early
+                    # The parent should break out of its loop in :recursive_phase
+                    parent_stage = :cleanup
+                elseif child_cycle ≠ 0 && child_cycle < parent_cycle
+                    # record the cycle will resolve at depth "cycle"
+                    parent_cycle = child_cycle
+                end
+
+                # Update both parent work state and result state
+                workspace.work_states[current_depth - 1] = VerifyMethodWorkState(parent_work.depth, parent_cause, parent_work.recursive_index, parent_stage)
+                workspace.result_states[current_depth - 1] = VerifyMethodResultState(parent_cycle, parent_minworld, parent_maxworld)
             end
+
+            # Pop current state
+            pop!(workspace.initial_states)
+            pop!(workspace.work_states)
+            pop!(workspace.result_states)
+            current_depth -= 1
         end
     end
-    if maxworld ≠ 0 && cycle ≠ depth
-        return cycle, minworld, maxworld
-    end
-    # If we are the top of the current cycle, now mark all other parts of
-    # our cycle with what we found.
-    # Or if we found a failed edge, also mark all of the other parts of the
-    # cycle as also having a failed edge.
-    while length(stack) ≥ depth
-        child = pop!(stack)
-        if maxworld ≠ 0
-            @atomic :monotonic child.min_world = minworld
-        end
-        @atomic :monotonic child.max_world = maxworld
-        if maxworld == validation_world && validation_world == get_world_counter()
-            Compiler.store_backedges(child, child.edges)
-        end
-        @assert visiting[child] == length(stack) + 1
-        delete!(visiting, child)
-        invalidations = _jl_debug_method_invalidation[]
-        if invalidations !== nothing && maxworld < validation_world
-            push!(invalidations, child, "verify_methods", cause)
-        end
-    end
-    return 0, minworld, maxworld
+
+    # Return results from the root call
+    return (root_result.child_cycle, root_result.result_minworld, root_result.result_maxworld)
```

Comment:
won't this still be an empty `VerifyMethodResultState(0, 0, 0)` ? `workspace.result_states[1]` may have been overwritten but `root_result` doesn't change

Reply:
Thanks for catching that. It used to be mutable and then I changed that without remembering I'd need to re-test. I merged, but happy for post-commit review too if anyone notices any other translations mistakes.